src/java.base/share/classes/java/io/Reader.java

```diff
@@ -457,7 +450,58 @@ private String readAllCharsAsString() throws IOException {
      * @since 25
      */
     public List<String> readAllLines() throws IOException {
-        return readAllCharsAsString().lines().toList();
+        List<String> lines = new ArrayList<>();
+        char[] cb = new char[1024];
+
+        int start = 0;
+        int pos = 0;
+        int limit = 0;
+        boolean skipLF = false;
+        int n;
+        while ((n = read(cb, pos, cb.length - pos)) != -1) {
+            limit = pos + n;
+            while (pos < limit) {
+                if (skipLF) {
+                    if (cb[pos] == '\n') {
+                        pos++;
+                        start++;
+                    }
+                    skipLF = false;
+                }
+                while (pos < limit) {
+                    char c = cb[pos++];
+                    if (c == '\n' || c == '\r') {
+                        lines.add(new String(cb, start, pos - 1 - start));
+                        skipLF = (c == '\r');
+                        start = pos;
+                        break;
+                    }
+                }
+                if (pos == limit) {
+                    int len = limit - start;
+                    if (len >= cb.length) {
```

Comment:
Observation: this algorithm will get less efficient as the line length approaches cb.length.
Starting at sb.length/2 It will take 2 reads for every line and will not resize the buffer.
I think resizing the buffer when the line length exceeds cb.length/2 will be more efficient (cpu wise).
The decoding of the bytes to chars and the setup of the decoder is the expensive part.
For most typical line length distributions it would not make a difference but would improve the performance of the outliners.

Reply:
> I think resizing the buffer when the line length exceeds cb.length/2 will be more efficient (cpu wise).

I concur. I missed that this had been changed in the most recent commit.

Reply:
So changed in [9eeed76](https://github.com/openjdk/jdk/pull/25863/commits/9eeed767ced6e7b86ac7c2c73fb2d32063d88a7a).